File: src/main/java/com/morph/engine/math/Matrix3f.java
package com.morph.engine.math;

public class Matrix3f {
	private float[] m;
	
	public Matrix3f(float...m) {
		this.m = m;
	}
	
	public String toString() {
		String result = "Matrix3f(";
		
		for (int i = 0; i < 8; i++) {
			result += m[i] + ", ";
			if (i % 3 == 2)
				result += "\n";
		}
		
		return result + m[8] + ")";
	}
	
	// TODO: BROKEN!
	public float getDeterminant() {
		float det = 0;
		
		for (int sign = 0; sign < 2; sign++) {
			for (int diagonal = 0; diagonal < 3; diagonal++) {
				float d = 1;
				for (int value = 0; value < 3; value++) {
					if (sign == 0) {
						d *= get((value + diagonal) % 3, value % 3);
					} else {
						d *= get(2 - ((value + diagonal) % 3), value % 3);
					}
				}
				
				if (sign == 0) {
					det += d;
				} else {
					det -= d;
				}
			}
		}
		
//		float det = 0;
//		
//		for (int i = 0; i < 3; i++) {
//			int sign = i % 2 == 0 ? 1 : -1;
//			det += sign * m[i] * minorDet(i, 0);
//		}
		
		return det;
	}
	
	public Matrix3f scale(float k) {
		float[] values = new float[9];
		for (int i = 0; i < 9; i++) {
			values[i] = m[i] * k;
		}
		
		return new Matrix3f(values);
	}
	
	public Matrix3f mulComp(Matrix3f mat) {
		float[] values = new float[9];
		for (int i = 0; i < 9; i++) {
			values[i] = m[i] * mat.get(i);
		}
		
		return new Matrix3f(values);
	}
	
	public float get(int i) {
		return m[i];
	}
	
	public float get(int i, int j) {
		return m[i + j * 3];
	}
	
	public float[] getValues() {
		return m;
	}
	
	public Matrix2f getMatrix2f(int a, int b, int c, int d) {
		return new Matrix2f(m[a], m[b], m[c], m[d]);
	}
	
	public Matrix3f getInverse() {
//		Matrix3f matOfMinors = new Matrix3f(
//				getMatrix2f(4, 5, 7, 8).getDeterminant(), getMatrix2f(3, 5, 6, 8).getDeterminant(), getMatrix2f(3, 4, 6, 7).getDeterminant(),
//				getMatrix2f(1, 2, 7, 8).getDeterminant(), getMatrix2f(0, 2, 6, 8).getDeterminant(), getMatrix2f(0, 1, 6, 7).getDeterminant(),
//				getMatrix2f(1, 2, 4, 5).getDeterminant(), getMatrix2f(0, 2, 3, 5).getDeterminant(), getMatrix2f(0, 1, 3, 4).getDeterminant()
//		);
		
		Matrix3f matOfMinors = new Matrix3f(
				minorDet(0, 0), minorDet(1, 0), minorDet(2, 0),
				minorDet(0, 1), minorDet(1, 1), minorDet(2, 1),
				minorDet(0, 2), minorDet(1, 2), minorDet(2, 2)
		);
		
		Matrix3f sign = new Matrix3f(
				1, -1, 1,
				-1, 1, -1,
				1, -1, 1
		);
		
		Matrix3f cofactor = matOfMinors.mulComp(sign);
		Matrix3f adjugate = cofactor.getTranspose();
		float invDet = 1f / getDeterminant();
		
		return adjugate.scale(invDet);
	}
	
	private Matrix2f minor(int i, int j) {
		float[] values = new float[4];
		int index = 0;
		for (int y = 0; y < 3; y++) {
			for (int x = 0; x < 3; x++) {
				if (x != i && y != j) {
					values[index] = get(x, y);
					index++;
				}
				if (index == 4) {
					break;
				}
			}
		}
		
		return new Matrix2f(values);
	}
	
	private float minorDet(int i, int j) {
		return minor(i, j).getDeterminant();
	}

	private Matrix3f getTranspose() {
		return new Matrix3f(
				m[0], m[3], m[6],
				m[1], m[4], m[7],
				m[2], m[5], m[8]
		);
	}
}

File: src/main/kotlin/com/morph/engine/math/Matrix3f.kt
package com.morph.engine.math

class Matrix3f(
        m00: Float, m01: Float, m02: Float,
        m10: Float, m11: Float, m12: Float,
        m20: Float, m21: Float, m22: Float
) {
    private val values: FloatArray = floatArrayOf(
            m00, m01, m02,
            m10, m11, m12,
            m20, m21, m22
    )

    val determinant: Float
        get() {
            val (reduction, swaps) = this.computeLUDecomposition()
            val power = (sequenceOf(1) + generateSequence { -1 }).take(swaps + 1).reduce {a, b -> a * b}
            return power * reduction[0, 0] * reduction[1, 1] * reduction[2, 2]
        }

    @Deprecated("Extremely slow O(n!) algorithm") val oldDeterminant: Float
        get() {
            var det = 0f

            for (sign in 0..1) {
                for (diagonal in 0..2) {
                    var d = 1f
                    for (value in 0..2) {
                        d *= if (sign == 0) {
                            get((value + diagonal) % 3, value % 3)
                        } else {
                            get(2 - (value + diagonal) % 3, value % 3)
                        }
                    }

                    if (sign == 0) {
                        det += d
                    } else {
                        det -= d
                    }
                }
            }

            return det
        }

    val inverse: Matrix3f
        get() {
            var result = Matrix3f(this.values)
            var inverse = Matrix3f.identity()

            fun swap(a: Int, b: Int) {
                result = result.swapRows(a, b)
                inverse = inverse.swapRows(a, b)
                println("Swapped rows $a and $b, result: $result, inverse: $inverse")
            }

            fun scale(a: Int, k: Float) {
                result = result.scaleRow(a, k)
                inverse = inverse.scaleRow(a, k)
                println("Scaled row $a by $k, result: $result, inverse: $inverse")
            }

            fun addScaled(a: Int, b: Int, k: Float) {
                result = result.addScaledRow(a, k, b)
                inverse = inverse.addScaledRow(a, k, b)
                println("Row operation on row $a using row $b by $k, result: $result, inverse: $inverse")
            }

            var row = 0
            var col = 0
            while (row <= 2 && col <= 2) {
                val max = (row..2).maxBy { Math.abs(result[it, col]) }!!
                if (result[max, col] == 0f) {
                    col++
                } else {
                    swap(row, max)
                    for (i in (row + 1)..2) {
                        val k = result[i, col] / result[row, col]
                        addScaled(i, row, -k)
                    }

                    row++
                    col++
                }
            }

            if (result[2, 2] != 0f) {
                addScaled(1, 2, -result[1, 2] / result[2, 2])
            }

            if (result[1, 1] != 0f) {
                addScaled(0, 1, -result[0, 1] / result[1, 1])
            }

            if (result[2, 2] != 0f) {
                addScaled(0, 2, -result[0, 2] / result[2, 2])
            }

            if (result[0, 0] != 0f) scale(0,  1 / result[0, 0])
            if (result[1, 1] != 0f) scale(1, 1 / result[1, 1])
            if (result[2, 2] != 0f) scale(2, 1 / result[2, 2])

            return inverse
        }

    constructor(m: FloatArray) : this(m[0], m[1], m[2], m[3], m[4], m[5], m[6], m[7], m[8])

    private val transpose: Matrix3f
        get() = Matrix3f(
                values[0], values[3], values[6],
                values[1], values[4], values[7],
                values[2], values[5], values[8]
        )

    override fun toString(): String {
        val result = StringBuilder("Matrix3f(")

        for (i in 0..7) {
            result.append(values[i]).append(", ")
            if (i % 3 == 2)
                result.append("\n")
        }

        return result.toString() + values[8] + ")"
    }

    @JvmName("scale")
    operator fun times(k: Float): Matrix3f {
        val values = FloatArray(9)
        for (i in 0..8) {
            values[i] = this.values[i] * k
        }

        return Matrix3f(values)
    }

    fun mulComp(mat: Matrix3f): Matrix3f {
        val values = FloatArray(9)
        for (i in 0..8) {
            values[i] = this.values[i] * mat[i]
        }

        return Matrix3f(values)
    }

    operator fun get(i: Int): Float {
        return values[i]
    }

    operator fun get(i: Int, j: Int): Float {
        return values[j + i * 3]
    }

    operator fun set(i: Int, j: Int, value: Float) {
        this.values[j + (i * 3)] = value
    }

    fun swapRows(a: Int, b: Int) : Matrix3f {
        require(a < 3 && b < 3) { "Attempt to swap rows outside 0-2" }
        val rows = values.asList().chunked(3).toMutableList()
        val temp = rows[a]
        rows[a] = rows[b]
        rows[b] = temp
        return Matrix3f(rows.flatten().toFloatArray())
    }

    fun scaleRow(a: Int, k: Float) : Matrix3f {
        val result = Matrix3f(this.values)
        result[a, 0] *= k
        result[a, 1] *= k
        result[a, 2] *= k
        return result
    }

    fun addScaledRow(unscaledRow: Int, scaleFactor: Float, scaledRow: Int) : Matrix3f {
        val rowAdded = Vector3f(this[scaledRow, 0], this[scaledRow, 1], this[scaledRow, 2]) * scaleFactor
        val rowKept = Vector3f(this[unscaledRow, 0], this[unscaledRow, 1], this[unscaledRow, 2])
        val newRow = rowAdded + rowKept
        val result = Matrix3f(this.values)
        result[unscaledRow, 0] = newRow.x
        result[unscaledRow, 1] = newRow.y
        result[unscaledRow, 2] = newRow.z
        return result
    }

    fun computeLUDecomposition() : Pair<Matrix3f, Int> {
        var result = this
        val pivot1 = sequenceOf(0, 1, 2).filter { result[it, 0] != 0f }.min()
        var swaps : Int = 0

        pivot1?.let {
            if (it != 0) {
                result = result.swapRows(it, 0)
                swaps++
            }
            for (i in 1..2) {
                if (result[i, 0] != 0f) {
                    result = result.addScaledRow(i, -result[i, 0] / result[0, 0], 0)
                }
            }
        }

        val pivot2 = sequenceOf(1, 2).filter { result[it, 1] != 0f }.min()
        pivot2?.let {
            if (it != 1) {
                result = result.swapRows(it, 1)
                swaps++
            }
            if (result[2, 1] != 0f) {
                result = result.addScaledRow(2, -result[2, 1] / result[1, 1], 1)
            }
        }

        return result to swaps
    }

    companion object {
        @JvmStatic fun identity() = Matrix3f(
                1f, 0f, 0f,
                0f, 1f, 0f,
                0f, 0f, 1f
        )

        @JvmStatic fun empty() = Matrix3f(
                0f, 0f, 0f,
                0f, 0f, 0f,
                0f, 0f, 0f
        )
    }
}
